src/query/datatype.rs

```diff
@@ -0,0 +1,253 @@
+            (DataType::List(l1), DataType::List(l2)) => Ok(l1 == l2),
+            (DataType::Dict(d1), DataType::Dict(d2)) => Ok(d1 == d2),
+            // We do not care about comparing functions
+            _ => Err(QueryError::InvalidType(format!("Cannot compare values of different types {:?} and {:?}", self, other))),
+        }
+    }
+}
+
+/* Required for query_eq when comparing two dicts */
+impl PartialEq for DataType {
+    fn eq(&self, other: &DataType) -> bool {
+        match (self, other) {
+            (DataType::None(), DataType::None()) => true,
+            // TODO: Comparisons of bool == num, bool == str
+            (DataType::Bool(b1), DataType::Bool(b2)) => b1 == b2,
+            (DataType::Number(n1), DataType::Number(n2)) => n1 == n2,
+            (DataType::String(s1), DataType::String(s2)) => s1 == s2,
+            (DataType::Event(e1), DataType::Event(e2)) => e1 == e2,
+            (DataType::List(l1), DataType::List(l2)) => l1 == l2,
+            (DataType::Dict(d1), DataType::Dict(d2)) => d1 == d2,
+            // We do not care about comparing functions
+            _ => false
+        }
+    }
+}
+
+impl TryFrom<&DataType> for Vec<DataType> {
+    type Error = QueryError;
+    fn try_from(value: &DataType) -> Result<Self, Self::Error> {
+        match value {
+            DataType::List(ref s) => Ok(s.clone()),
+            ref invalid_type => Err(QueryError::InvalidFunctionParameters(
+                format!("Expected function parameter of type List, got {:?}", invalid_type)
+            ))
+        }
+    }
+}
+
+impl TryFrom<&DataType> for String {
+    type Error = QueryError;
+    fn try_from(value: &DataType) -> Result<Self, Self::Error> {
+        match value {
+            DataType::String(s) => Ok(s.clone()),
+            ref invalid_type => Err(QueryError::InvalidFunctionParameters(
+                format!("Expected function parameter of type List of Strings, list contains {:?}", invalid_type)
+            ))
+        }
+    }
+}
+
+impl TryFrom<&DataType> for Vec<String> {
+    type Error = QueryError;
+    fn try_from(value: &DataType) -> Result<Self, Self::Error> {
+        let mut tagged_strings: Vec<DataType> = value.try_into()?;
+        let mut strings = Vec::new();
+        for string in tagged_strings.drain(..) {
+            let s: String = (&string).try_into()?;
+            strings.push(s);
+        }
+        return Ok(strings);
+    }
+}
+
+impl TryFrom<&DataType> for Rule {
+    type Error = QueryError;
+    fn try_from(value: &DataType) -> Result<Self, Self::Error> {
+        let rulemap: HashMap<String, String> = (match value {
+            DataType::Dict(d) => {
+                let map: HashMap<String, String> = d.iter().map(|(k, v)| {
+                    let s: String = v.try_into().unwrap();
+                    (k.clone(), s.clone())
+                }).collect();
+                Ok(map)
+            },
+            DataType::String(s) => {
```

Comment:
I can't decide if it is a good idea to accept strings and assume that they are a regex.
It feels like it's a nice way to easily use the API, but at the same time I want things to be "explicit rather than implicit".

Reply:
This will be removed, almost forgot about it. Thanks!